### PR TITLE
Basic functionality

### PR DESCRIPTION
**Pull request by anetso18**

### CODE REVIEW

File: WormsApp/MapperWrapper.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#ifndef MAPPERWRAPPER_H
+#define MAPPERWRAPPER_H
```
Comment (pokorska-work) [inline WormsApp/MapperWrapper.h]: It is recommended to use `#pragma once` instead.

File: WormsApp/MapperWrapper.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+#include <unordered_map>
+#include <utility>
+#include<thread>
+
+template <typename KeyType, typename valueType> class MapperWrapper
```
Comment (pokorska-work) [inline WormsApp/MapperWrapper.h]: We usually format it like this:
```
template <typename KeyType, typename valueType>
class MapperWrapper
```
This makes it easier to read the class name (it's not far away to the right).

File: WormsApp/MapperWrapper.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public:
+        void add(KeyType key, valueType&& value)
+        {
+        mapper.emplace(key, std::forward<std::thread>(value));
```
Comment (pokorska-work) [inline WormsApp/MapperWrapper.h]: I think we force in this line the `valueType` to become `std::thread`, so there is no point in having `typename valueType` in the class template.

I think we should either keep it all generic or keep it all fully typed. I'd be ok with templating only the key of the map, but it would look a bit weird.

File: WormsApp/board.cpp [hunk 4]
```diff
@@ -1,26 +1,54 @@
...
 
+Board::~Board() {
+ for(auto& w: getWorms()){
+     w.second.join();}
+ }
+
+
 void Board::addWorm(WormType type, int x, int y) {
-  // TODO: Implement the body of the function. It should add a worm
+  // TODO:   the body of the function. It should add a worm
```
Comment (pokorska-work) [inline WormsApp/board.cpp]: Whoops :)

File: WormsApp/board.cpp [hunk 5]
```diff
@@ -1,26 +1,54 @@
...
   // call operator - operator()(int) - defined in Worm class.
   // Note: there are a few class members (like wormTypes_) that need to be
   // updated as well to have full information about the worm that is being
   // added. They are usually based on worm's id which should be uniquely given
   // here.
+     board_[x][y] = nextId_;
+    if (type == Lazy)
+    {
+        LazyWorm w = LazyWorm(x,y,this);
+        worms_.add(nextId_,std::thread(std::move(w),nextId_));
```
Comment (pokorska-work) [inline WormsApp/board.cpp:33]: I recently discovered that this creates and instance of a `std::thread` and then copies it when inserting to the map. In general this may be dangerous depending on the implementation of a copy constructor and sometimes may not compile (when copy constructor is not defined). Fortunately thread is implemented wisely and will take care of it - copy constructor as well as assignment operator are moving the thread to the new one. Inside, new thread is created and immediately joined with the old one.

TLDR: It's ok, you just need to be careful in the future.

File: WormsApp/board.cpp [hunk 6]
```diff
@@ -1,26 +1,54 @@
...
 }
 
 void Board::update(int id, int oldX, int oldY, int newX, int newY) {
+  std::lock_guard<std::mutex> lock(mut);
+  if(checkKill(id))
+       return;
   int targetId = board_[newX][newY];
   if (targetId != 0 && targetId != id)
-    killed_.insert(targetId);
+  killed_.insert(targetId);
```
Comment (pokorska-work) [inline WormsApp/board.cpp]: It goes under `if`, so it should be indented.

File: WormsApp/makefile [hunk 7]
```diff
@@ -1,2 +1,2 @@
 all:
-	g++ -std=c++14 -o wormApp.e main.cpp graphics.cpp worm.cpp board.cpp `pkg-config gtkmm-3.0 --cflags --libs` -Wall -Wextra -Werror
+	g++ -std=c++14 -o wormApp.e main.cpp graphics.cpp worm.cpp board.cpp `pkg-config gtkmm-3.0 --cflags --libs` #-Wall -Wextra -Werror
```
Comment (pokorska-work) [inline WormsApp/makefile]: These flags are useful. They usually point some minor issues, but sometimes it may catch something you wouldn't notice.

File: WormsApp/worm.cpp [hunk 8]
```diff
@@ -14,12 +15,34 @@ using std::unordered_set;
...
 
-  // TODO: Implement the body of the function. It should perform one step
-  // based on currDir_ and modify class variables x_ and y_.
+   //TODO: Implement the body of the function. It should perform one step
+   //based on currDir_ and modify class variables x_ and y_.
+
+  switch (currDir_)
+  {
+  case LEFT:
+      (x_>0)?--x_:x_%=board_->getWidth();
```
Comment (pokorska-work) [inline WormsApp/worm.cpp]: It looks like obfuscated code :) Please consider putting some spaces to make it readable.

File: WormsApp/worm.cpp [hunk 9]
```diff
@@ -43,7 +66,7 @@ void LazyWorm::run() {
       (int)currDir_--;
       if (currDir_ < 0) currDir_ += 4;
     }
-    this->move();
+     this->move();
```
Comment (pokorska-work) [inline WormsApp/worm.cpp:75]: I think you didn't mean to add this space.

File: WormsApp/worm.cpp [hunk 10]
```diff
@@ -14,12 +15,34 @@ using std::unordered_set;
 using std::hash;
 using std::vector;
 
+std::mutex mut;
```
Comment (pokorska-work) [inline WormsApp/worm.cpp]: Some leftover from previous idea?

File: WormsApp/MapperWrapper.h [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+        }
+
+        std::unordered_map<KeyType, valueType> mapper;
```
Comment (pokorska-work) [inline WormsApp/MapperWrapper.h]: This is a public variable of a class. Usually you want to hide your internal variables so that no one else from outside can modify (and possibly brake) them. If you want to access such variable you create a getter function that returns a constant reference to such variable or creates a copy and returns it.

Returning const reference (another const after the method name means we don't modify our class):
```
const std::unordered_map<KeyType, valueType>& getMapper() const {
  return mapper;
}
```

Copy the map (not recommended - it's time consuming and may mess up our threads):
```
std::unordered_map<KeyType, valueType> getMapper() const {
  return mapper;
}
```

File: WormsApp/MapperWrapper.h (hunk 2 above)
Comment (pokorska-work) [inline WormsApp/MapperWrapper.h]: nit: names should follow one style - `KeyType` and `ValueType` or alternatively `keyType` and `valueType`.

File: WormsApp/board.h [hunk 12]
```diff
@@ -13,15 +14,23 @@ enum WormType {
...
+
 
- public:
+public:
   Board(int width, int height);
-  virtual ~Board() { }
+
+  std::unordered_map<int, std::thread> getWorms()
+  {
+      return std::move (worms_.mapper);
```
Comment (pokorska-work) [inline WormsApp/board.h]: I don't understand this - is it needed? You destroy the mapper inside `worms_` while whole board class can still live. I think it may cause problems.

**Review by pokorska-work — CHANGES_REQUESTED** [review]
Good work!

File: WormsApp/MapperWrapper.h (hunk 1 above)
Comment (anetso18) [reply WormsApp/MapperWrapper.h]: Changed)

File: WormsApp/MapperWrapper.h (hunk 2 above)
Comment (anetso18) [reply WormsApp/MapperWrapper.h]: Done.

File: WormsApp/MapperWrapper.h (hunk 3 above)
Comment (anetso18) [reply WormsApp/MapperWrapper.h]: Done.

File: WormsApp/board.cpp (hunk 5 above)
Comment (anetso18) [reply WormsApp/board.cpp:33]: Ok. I see)

File: WormsApp/MapperWrapper.h (hunk 11 above)
Comment (anetso18) [reply WormsApp/MapperWrapper.h]: I know that const reference should be used in getters,  but I had a troubles with that when joining the threads inside a mapper. So, I made It without const .

File: WormsApp/board.h (hunk 12 above)
Comment (anetso18) [reply WormsApp/board.h]: Removed, getMapper() method is using instead.

File: WormsApp/worm.cpp (hunk 10 above)
Comment (anetso18) [reply WormsApp/worm.cpp]: Removed

File: WormsApp/worm.cpp (hunk 8 above)
Comment (anetso18) [reply WormsApp/worm.cpp]: Done.

File: WormsApp/makefile (hunk 7 above)
Comment (anetso18) [reply WormsApp/makefile]: Uncommented

File: WormsApp/worm.cpp [hunk 13]
```diff
@@ -28,8 +26,12 @@ void Worm::move() {
   switch (currDir_)
   {
   case LEFT:
-      (x_>0)?--x_:x_%=board_->getWidth();
```
Comment (pokorska-work) [inline WormsApp/worm.cpp]: I actually thought about adding some spaces: `x_ > 0 ? --x_ : x_ %= board_->getWidth();`, but with `if-else` is even better I think.

File: WormsApp/worm.cpp [hunk 14]
```diff
@@ -28,8 +26,12 @@ void Worm::move() {
...
+      if(x_>0)
```
Comment (pokorska-work) [inline WormsApp/worm.cpp:29]: How about spaces here? `if(x_ > 0)` or `if (x_ > 0)`?

File: WormsApp/worm.cpp [hunk 15]
```diff
@@ -28,8 +26,12 @@ void Worm::move() {
...
+          --x_;
+      else
+          x_%=board_->getWidth();
```
Comment (pokorska-work) [inline WormsApp/worm.cpp:32]: This else means that `x_` is negative and I'm not sure whether making modulo (`%`) always makes it positive. Instead I'd do `x_ += board_->getWidth()`.

File: WormsApp/worm.cpp [hunk 16]
```diff
@@ -38,9 +40,13 @@ void Worm::move() {
...
       y_%= board_->getHeight();
-break;
+      break;
+
   case DOWN:
-       (y_>0)?--y_:y_%=board_->getHeight();
+       if(y_>0)
+           --y_;
+       else
+           y_%=board_->getHeight();
```
Comment (pokorska-work) [inline WormsApp/worm.cpp:49]: Same here.

File: WormsApp/board.cpp [hunk 17]
```diff
@@ -14,13 +13,13 @@ Board::Board(int width, int height) : nextId_{1}, killAll_{false} {
 }
 
 Board::~Board() {
- for(auto& w: getWorms()){
-     w.second.join();}
- }
-
+    for(auto& m: std::move(worms_.getMapper())){
```
Comment (pokorska-work) [inline WormsApp/board.cpp:16]: I wouldn't do `std::move` here. I'm not sure how these objects will be destroyed because of this. Your wrapper has private map which will be destroyed when the wrapper is destroyed, while here you are stealing its reference and you are invalidating it with `std::move`.

I'd stick with:
```
for (auto& m : worms_.getMapper()) {
  m.second.join();
}
```

I'd also do `getMapper` method return a const reference, but probably compiler will not allow for that. You may give it a try.

File: WormsApp/board.cpp [hunk 18]
```diff
@@ -48,7 +47,7 @@ void Board::update(int id, int oldX, int oldY, int newX, int newY) {
        return;
   int targetId = board_[newX][newY];
   if (targetId != 0 && targetId != id)
-  killed_.insert(targetId);
+   killed_.insert(targetId);
```
Comment (pokorska-work) [inline WormsApp/board.cpp:50]: You should keep indentation consistent - I used two-space indentation, you seem to be using four-space indentation and here you inserted only a single space.

File: WormsApp/MapperWrapper.h [hunk 19]
```diff
@@ -1,18 +1,22 @@
...
+template <typename keyType, typename valueType>
+class MapperWrapper
 {
-    public:
-        void add(KeyType key, valueType&& value)
-        {
-        mapper.emplace(key, std::forward<std::thread>(value));
-        }
+public:
+    void add(keyType key, valueType&& value)
```
Comment (pokorska-work) [inline WormsApp/MapperWrapper.h:11]: This is great that you used lvalue reference here, because we actually want to move objects into our map, so it is the most appropriate type.

**Review by pokorska-work — CHANGES_REQUESTED** [review]
